doc/continuous_integration/dockerfile.md

```diff
@@ -0,0 +1,120 @@
+# Dockerfile Documentation
+
+Dockerfiles, denoted via the `(<name>.)Dockerfile` file name extension, provides repeatable and reproducible means to build and test the project, as well as build images for running container based CI services. Further references on writing and building Dockerfiles, such as syntax and tooling can be found here:
+
+* [Dockerfile reference](https://docs.docker.com/engine/reference/builder)
+* [Best practices for writing Dockerfiles](https://docs.docker.com/develop/develop-images/dockerfile_best-practices)
+
+The Dockerfiles for this project are built upon parent images from upstream repos on DockerHub, thus abbreviating environmental setup and build time, yet written in a parameterized style to remain ROS2 distro agnostic. This keeps them easily generalizable for future ROS2 releases or for switching between custom parent images. When choosing the parent image, a tradeoff may persist between choosing a larger tag with more than what you need pre-installed, saving time building the image locally, vs. choosing a smaller tag without anything you don't need, saving time pulling or pushing the image remotely. Given the use of multiple build stages, they're consequently best approached by reading from top to bottom in the order in which image layers are appended. More info on upstream repos on DockerHub can be found here:
+
+* [ROS Docker Images](https://hub.docker.com/_/ros)
+  * DockerHub repo for official images
+* [ROS Dockerfiles](https://github.com/osrf/docker_images)
+  * GitHub repo for OSRF Dockerfiles
+* [Official Images on Docker Hub](https://docs.docker.com/docker-hub/official_images)
+
+While the main [`Dockerfile`](/Dockerfile) at the root of the repo is used for development and continuous integration, the [`.dockerhub/`](/.dockerhub) directory contains additional Dockerfiles that can be used for building the project entirely from scratch, include the minimal spanning set of recursive ROS2 dependencies from source, or building the project from a released ROS2 distro using available pre-built binary dependencies. These are particularly helpful for developers needing to build/test the project using a custom ROS2 branch, or for a users building with an alternate ROS2 base image, but are not used for the CI pipeline. We'll walk through the main Dockerfile here, although all of them follow the same basic pattern.
```

Comment:
> Please explain to me then why we have them at all.

In the same paragraph:

> These are particularly helpful for developers needing to build/test the project using a custom ROS2 branch, or for a users building with an alternate ROS2 base image

They are also referenced under the Buildkit and Advanced sections respectively, serving as examples for building nav2 without using the ros2 master nightly build images, demoing optimizations DockerHub doesn't yet support, and thus un-implemented in the main Dockerfile.

> A difference for other Dockerfiles, not needing to be built by DockerHub and thus not limited in backwards compatibility, is the use of newer mount syntax options in buildkit, allowing for persistent caching of downloaded apt packages and ccache files across successful builds of the same Dockerfile.

Reply:
I read the buildkit stuff after this, so that makes sense to keep one of them around for.

On the 'useful for other needs' like custom ros2 branches / alternative bases, I don't think that should be on the Nav2 project to support that. If someone was going to do something like that, I don't think anyone's first instinct would be to look at the navigation metapackage for a dockerfile that does that. I would think that's a bit out of scope for the project needs and what people would look to nav2 for. However, I'd be OK leaving them if we added a comment block on top for those 2 specifically laying out what it does and what its for separately from CI (and making clear that these are not used anywhere in the CI or stack currently). 

The aim is to make it clear what's in use, where and what's not and why.